src/bytecode/concrete.py

```diff
@@ -1223,34 +1268,60 @@ def concrete_instructions(self) -> None:
 
             assert isinstance(instr, Instr)
 
-            if instr.lineno is not UNSET and instr.lineno is not None:
-                lineno = instr.lineno
-            elif instr.lineno is UNSET:
-                instr.lineno = lineno
+            if instr.location is not UNSET and instr.location is not None:
+                location = instr.location
 
+            instr_name = instr.name
+            opcode = instr._opcode
             arg = instr.arg
             is_jump = False
             if isinstance(arg, Label):
                 label = arg
                 # fake value, real value is set in compute_jumps()
-                arg = 0
+                c_arg = 0
                 is_jump = True
-            elif instr.opcode in _opcode.hasconst:
-                arg = self.add_const(arg)
-            elif instr.opcode in _opcode.haslocal:
-                assert isinstance(arg, str)
-                arg = self.add(self.varnames, arg)
-            elif instr.opcode in _opcode.hasname:
-                if instr.name in BITFLAG_INSTRUCTIONS:
+            elif opcode in _opcode.hasconst:
+                c_arg = self.add_const(arg)
+            elif opcode in _opcode.haslocal:
+                if opcode in DUAL_ARG_OPCODES:
+                    assert (
+                        isinstance(arg, tuple)
+                        and len(arg) == 2
+                        and isinstance(arg[0], str)
+                        and isinstance(arg[1], str)
+                    )
+                    arg1_index = self.add(self.varnames, arg[0])
+                    arg2_index = self.add(self.varnames, arg[1])
+                    if arg1_index > 16 or arg2_index > 16:
+                        parts = instr.name.split("_")
+                        n1 = "_".join(parts[:2])
+                        n2 = "_".join(parts[2:])
```

Comment:
Whilst currently correct, this logic feels a bit brittle. Perhaps it would be safer to have a mapping between dual arg opcodes and the pair of single instructions? E.g.
```python
DUAL_ARG_OPCODES_SINGLE_OPS = {
    "LOAD_FAST_LOAD_FAST": ("LOAD_FAST", "LOAD_FAST"),
    ...,
}

Reply:
I can do that yes. (Hopefully sometimes next week)